subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/WatchableHierarchies.java

```diff
@@ -0,0 +1,133 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.watch.registry.impl;
+
+import org.gradle.internal.file.DefaultFileHierarchySet;
+import org.gradle.internal.file.FileHierarchySet;
+import org.gradle.internal.file.FileMetadata;
+import org.gradle.internal.snapshot.CompleteDirectorySnapshot;
+import org.gradle.internal.snapshot.CompleteFileSystemLocationSnapshot;
+import org.gradle.internal.snapshot.FileSystemSnapshotVisitor;
+import org.gradle.internal.snapshot.SnapshotHierarchy;
+
+import javax.annotation.CheckReturnValue;
+import java.io.File;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+public class WatchableHierarchies {
+    private final Predicate<String> watchFilter;
+
+    private FileHierarchySet watchableHierarchies = DefaultFileHierarchySet.of();
+
+    public WatchableHierarchies(Predicate<String> watchFilter) {
+        this.watchFilter = watchFilter;
+    }
+
+    public void registerWatchableHierarchy(File watchableHierarchy, SnapshotHierarchy root) {
+        String watchableHierarchyAbsolutePath = watchableHierarchy.getAbsolutePath();
+        if (!watchFilter.test(watchableHierarchyAbsolutePath)) {
+            throw new RuntimeException(String.format(
+                "Unable to watch directory '%s' since it is within Gradle's caches",
+                watchableHierarchyAbsolutePath
+            ));
+        }
+        if (!watchableHierarchies.contains(watchableHierarchyAbsolutePath)) {
+            checkThatNothingExistsInNewWatchableHierarchy(watchableHierarchyAbsolutePath, root);
+        }
+        watchableHierarchies = watchableHierarchies.plus(watchableHierarchy);
+    }
+
+    @CheckReturnValue
+    public SnapshotHierarchy removeUnwatchedSnapshots(Stream<File> watchedHierarchies, SnapshotHierarchy root, Invalidator invalidator) {
+        this.watchableHierarchies = DefaultFileHierarchySet.of(watchedHierarchies::iterator);
+        RemoveUnwatchedFiles removeUnwatchedFilesVisitor = new RemoveUnwatchedFiles(root, invalidator);
+        root.visitSnapshotRoots(snapshotRoot -> snapshotRoot.accept(removeUnwatchedFilesVisitor));
+        return removeUnwatchedFilesVisitor.getRootWithUnwatchedFilesRemoved();
+    }
+
+    private void checkThatNothingExistsInNewWatchableHierarchy(String watchableHierarchy, SnapshotHierarchy root) {
+        root.visitSnapshotRoots(watchableHierarchy, snapshotRoot -> {
+            if (!isInWatchableHierarchy(snapshotRoot.getAbsolutePath()) && !ignoredForWatching(snapshotRoot)) {
+                throw new RuntimeException(String.format(
```

Comment:
I wonder if it would make sense to have a `FileWatchingException` to use instead of `RuntimeException`. Same for the one thrown in `registerWatchableHierarchy()` above.

Reply:
Good question. We already have `WatchingNotSupportedException`, which we treat slightly different since we don't print the stack trace. For the two validations we do in `WatchableHierarchies`, it could be better to use `IllegalStateException`, since this shouldn't show up in user builds, right? Or maybe using a generic `FileWatchingException` would be best here. WDYT?

Reply:
I would be happy with an `IllegalStateException`. It carries more relevant information than `RuntimeException`.

Reply:
Using `IllegalStateException` now.